sidecar/src/service/serialized_tracer_header_tags.rs

```diff
@@ -0,0 +1,163 @@
+// Copyright 2021-Present Datadog, Inc. https://www.datadoghq.com/
+// SPDX-License-Identifier: Apache-2.0
+
+use datadog_trace_utils::trace_utils::TracerHeaderTags;
+use serde::{Deserialize, Serialize};
+use std::convert::TryFrom;
+use std::io;
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct SerializedTracerHeaderTags {
+    data: Vec<u8>,
+}
+
+/// `TryFrom` trait implementation for converting from `SerializedTracerHeaderTags` to
+/// `TracerHeaderTags`.
+///
+/// # Errors
+///
+/// Returns an `io::Error` if the deserialization of the `SerializedTracerHeaderTags` data fails.
+///
+/// # Examples
+///
+/// ```
+/// use bincode;
+/// use datadog_sidecar::service::SerializedTracerHeaderTags;
+/// use datadog_trace_utils::trace_utils::TracerHeaderTags;
+/// use std::convert::TryInto;
+///
+/// let tracer_header_tags = TracerHeaderTags {
+///     lang: "Rust",
+///     lang_version: "1.55.0",
+///     lang_interpreter: "rustc",
+///     lang_vendor: "Mozilla",
+///     tracer_version: "0.1.0",
+///     container_id: "1234567890",
+///     client_computed_top_level: true,
+///     client_computed_stats: false,
+/// };
+///
+/// let serialized: SerializedTracerHeaderTags = tracer_header_tags.try_into().unwrap();
+///
+/// let result: Result<TracerHeaderTags, _> = (&serialized).try_into();
+/// assert!(result.is_ok());
+/// ```
+impl<'a> TryFrom<&'a SerializedTracerHeaderTags> for TracerHeaderTags<'a> {
+    type Error = io::Error;
+
+    fn try_from(serialized: &'a SerializedTracerHeaderTags) -> Result<Self, Self::Error> {
+        bincode::deserialize(serialized.data.as_slice())
+            .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))
+    }
+}
+
+/// `TryFrom` trait implementation for converting from `TracerHeaderTags` to
+/// `SerializedTracerHeaderTags`.
+///
+/// # Errors
+///
+/// Returns a `bincode::Error` if the serialization of the `TracerHeaderTags` data fails.
+///
+/// # Examples
+///
+/// ```
+/// use datadog_sidecar::service::SerializedTracerHeaderTags;
+/// use datadog_trace_utils::trace_utils::TracerHeaderTags;
+/// use std::convert::TryInto;
+///
+/// let tracer_header_tags = TracerHeaderTags {
+///     lang: "Rust",
+///     lang_version: "1.55.0",
+///     lang_interpreter: "rustc",
+///     lang_vendor: "Mozilla",
+///     tracer_version: "0.1.0",
+///     container_id: "1234567890",
+///     client_computed_top_level: true,
+///     client_computed_stats: false,
+/// };
+///
+/// let serialized: Result<SerializedTracerHeaderTags, _> = tracer_header_tags.try_into();
+/// assert!(serialized.is_ok());
+/// ```
+// DEV: This implementation is not tested for the case where bincode raises an error because there
+// is no reasonable way to force bincode to fail during serialization without modifying the
+// `TracerHeaderTags` struct.
+impl<'a> TryFrom<TracerHeaderTags<'a>> for SerializedTracerHeaderTags {
```

Comment:
bincode is infallible for Strings. That's why the code did not use TryFrom, but just From. This was a deliberate choice, not unsafety.

Just as you note, there's no reasonable way for it to fail.

Reply:
For deserialization it's probably okay to check.